on_device_training/mobile/ios/MyVoice/Trainer.swift

```diff
@@ -0,0 +1,125 @@
+import Foundation
+import AVFoundation
+import onnxruntime_training_objc
+
+class Trainer {
+    private let ortEnv: ORTEnv
+    private let trainingSession: ORTTrainingSession
+    private let checkpoint: ORTCheckpoint
+    private var recordingCounter: Int
+    private let sampleAudioRecordings: Int
+    private let kEpoch: Int = 3
+    
+    enum TrainerError: Error {
+        case Error(_ message: String)
+    }
+    
+    init(sampleAudioRecordings : Int) throws {
+        ortEnv = try ORTEnv(loggingLevel: ORTLoggingLevel.warning)
+        
+        // get path for artifacts
+        guard let trainingModelPath = Bundle.main.path(forResource: "training_model", ofType: "onnx") else {
+            throw TrainerError.Error("Failed to find training model file.")
+        }
+        
+        guard let evalModelPath = Bundle.main.path(forResource: "eval_model",ofType: "onnx") else {
+            throw TrainerError.Error("Failed to find eval model file.")
+        }
+        
+        guard let optimizerPath = Bundle.main.path(forResource: "optimizer_model", ofType: "onnx") else {
+            throw TrainerError.Error("Failed to find optimizer model file.")
+        }
+        
+        guard let checkpointPath = Bundle.main.path(forResource: "checkpoint", ofType: nil) else {
+            throw TrainerError.Error("Failed to find checkpoint file.")
+        }
+        
+        checkpoint = try ORTCheckpoint(path: checkpointPath)
+        
+        trainingSession = try ORTTrainingSession(env: ortEnv, sessionOptions: ORTSessionOptions(), checkpoint: checkpoint, trainModelPath: trainingModelPath, evalModelPath: evalModelPath, optimizerModelPath: optimizerPath)
+        
+        self.sampleAudioRecordings = sampleAudioRecordings
+        recordingCounter = 0
+    }
+    
+    func exportModelForInference() throws {
+        guard let libraryDirectory = FileManager.default.urls(for: .libraryDirectory, in: .userDomainMask).first else {
+            throw TrainerError.Error("Failed to find library directory ")
+        }
+        
+        let modelPath = libraryDirectory.appendingPathComponent("inference_model.onnx").path
+        try trainingSession.exportModelForInference(withOutputPath: modelPath, graphOutputNames: ["output"])
+    }
+    
+    func train(audio: Data)  -> Result<Void, Error> {
+        return Result<Void, Error> { ()  in
+            for _ in 0..<kEpoch {
```

Comment:
I think the conventional meaning of "epoch" is an iteration through all the training data.
https://developers.google.com/machine-learning/glossary/#epoch

here we just train each sample consecutively `kEpoch` times.

I'm not too sure about this approach from a machine learning standpoint. @baijumeswani and @askhade what do you think?

we could also save the user-recorded files first and train after they are all available.

Reply:
We should not train the same sample consecutively epoch times. 

In this demo, given we let the user speak multiple times and use the data to train the model, I suggest we {train -> optimize -> reset grad} for every iteration. Basically, keep the global batch size to 1. We do need to test the accuracy with this approach. 

Another approach is to train for all the inputs that user provides and once all the user inputs are used we run the optimize step. Not sure how easy or difficult this is to do from the UI.

Reply:
Just checked the trainStep method and we already have global batch size 1. In that case, we can simply remove the consecutive training for epock times. Just train once with given data.

Reply:
From my testing earlier (on the image classification example), just training once on the sample might not be enough and will potentially not yield good results. You could try this out in python to see if that works.

If not, I think @edgchen1's suggestion of storing the file locally until the training is complete might be a good idea. Also, another thing to keep in mind might be to reorder the input data on every epoch if possible.

Reply:
I have changed training flow, so it performs training at the very end after all training recording is done. I am also shuffling data at every epoch